keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1300 @@
+
+The resources (mounts, devices, CDI devices, Kubernetes resources) in the
+CreateContainer request should be identical to what was (pre-)informed in the
+RunPodSandbox request. If they are different, the CRI runtime may fail the
+container creation, for example because changes cannot be applied after a
+VM-based Pod has been created.
+
+#### UpdateContainerResourcesRequest
+
+The UpdateContainerResourcesRequest message is extended to pass down unmodified
+resource requests from the PodSpec.
+
+```diff
+ message UpdateContainerResourcesRequest {
+     // ID of the container to update.
+     string container_id = 1;
+     // Resource configuration specific to Linux containers.
+     LinuxContainerResources linux = 2;
+     // Resource configuration specific to Windows containers.
+     WindowsContainerResources windows = 3;
+     // Unstructured key-value map holding arbitrary additional information for
+     // container resources updating. This can be used for specifying experimental
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 5;
+ }
+```
+
+Note that mounts, devices, CDI devices are not part of the
+UpdateContainerResourcesRequest message and this proposal does not suggest
+adding them.
+
+#### UpdatePodSandboxResources
+
+The In-Place Update of Pod Resources ([KEP-1287][kep-1287]) Beta
+([PR][kep-1287-beta-pr]) proposes to add new UpdatePodSandboxResources rpc to
+inform the CRI runtime about the changes in the pod resources.
+
+The UpdatePodSandboxResourcesRequest message is extended similarly to the
+[PodSandboxConfig](#podsandboxconfig) message to contain information about
+resources of all its containers. In UpdatePodSandboxResourcesRequest this will
+reflect the updated resource requirements of the containers.
+
+```diff
+ message UpdatePodSandboxResourcesRequest {
+     // ID of the PodSandbox to update.
+     string pod_sandbox_id = 1;
+ 
+     // Optional overhead represents the overheads associated with this sandbox
+     LinuxContainerResources overhead = 2;
+     // Optional resources represents the sum of container resources for this sandbox
+     LinuxContainerResources resources = 3;
+ 
+     // Unstructured key-value map holding arbitrary additional information for
+     // sandbox resources updating. This can be used for specifying experimental
+     // resources to update or other options to use when updating the sandbox.
+     map<string, string> annotations = 4;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 5;
+ }
+```
+
+The implementation will be synced with [KEP-1287][kep-1287].
+
+### kubelet
+
+Kubelet code is refactored/modified so that all container resources are known
+before sandbox creation. This mainly consists of preparing all mounts (of all
```

Comment:
Is there a possibility that today some mounts can ONLY be created when container is being created (some thing that init container needs to initialize). Will this break this behavior?

Reply:
This was discussed/investigated in https://github.com/kubernetes/enhancements/pull/4113#discussion_r1478654403

I think there should breakages because of this.

WDYT @SergeyKanzhelev ?

Reply:
When appropriate image mounts included in the container config are added at create container time when we build the rootfs.. additionally oci level plugins, nri plugins, cdi, and cri proxy tools may add mounts. But these are not the "extra" mounts passed in from the host level CRI client (kubelet), additionally the OCI image volume mount is added by the container runtime during create container. In the OCI volume case this is why I'm asking for the oci volume image reference..

Reply:
The `Mount` structure is exactly the same (with the same information) that what will be passed as part of the CreateContainerRequest/ContainerConfig message

Reply:
that works..

Reply:
it actually goes back to this comment: https://github.com/kubernetes/enhancements/pull/4113#discussion_r1468237753

Would it be expected that runtimes will need to mount everything at pod sandbox creation time now? Or the behavior is not defined? Will we require that the old way (without passing all this extra context in sandbox creation) should continue working?

Reply:
The runtimes do not need to change the way they operate. They don't need to mount everything at sandbox creation.

All of the data is something that the runtime CAN use if it needs it. E.g. [this paragraph](https://github.com/kubernetes/enhancements/blob/ea204299c71c4571ff0f38fa51e589e9104ccb16/keps/sig-node/4112-passdown-resources-to-cri/README.md#:~:text=The%20container%20runtime%20may,with%20the%20CreateContainer%20request.) of the proposal tries to state that. Maybe we need to state that more clearly in more places(?) And certainly comments in the cri proto file.

Reply:
The CAN part worries me. If something CAN be used the only assumption we can make is that it IS BEING USED. This is why maybe we should have two modes of operations

Reply:
@SergeyKanzhelev as it was mentioned above, this information is really needed in some scenarios, and OCI runtimes that are dependent on that information will be using it, to fix current problem. Other runtimes that don't have problems now are expected to ignore it.